### DISCUSSION TITLE
How do I use vcpkg together with Visual Studio 2019 and CUDA 11.3.1 ?

### DISCUSSION THREAD

**Discussion by EajksEajks** [microsoft/vcpkg#18697] [Q&A]
Hi

I was using cmake together with VStudio2019 and CUDA 11.3.1 and it worked fine. Because I would like to use additional C++ packages, I installed the vcpkg toolchain. Unfortunately, it doesn't find CUDA anymore!!! If I remove the vcpkg toolchain it finds it again. So what am I doing wrong? Any help is welcome.

The error message I get:
CMake Error at C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.20/Modules/CMakeDetermineCUDACompiler.cmake:207 (message):  
  Couldn't find CUDA library root.

The cmake command line
cmake ^
    -G "NMake Makefiles" ^
    -D CMAKE_C_COMPILER=cl ^
    -D CMAKE_CXX_COMPILER=cl ^
    -D CMAKE_BUILD_TYPE=Release ^
    -D CMAKE_TOOLCHAIN_FILE=C:\Tools\VCPKG\scripts\buildsystems\vcpkg.cmake ^
    -D PYBIND11_PYTHON_VERSION=3.7 ^
    ..\..\..\Components

The list of C++ packages I installed with vcpkg:
./vcpkg install caf:x64-windows
./vcpkg install cli11:x64-windows
./vcpkg install gtest:x64-windows
./vcpkg install pybind11:x64-windows
./vcpkg install spdlog:x64-windows
./vcpkg install cuda:x64-windows

**Comment by PhoebeHui** [comment]
@EajksEajks, could you provide us more information, like repro steps, repro code?

looks similar issue with https://github.com/microsoft/vcpkg/issues/14419

**Comment by EajksEajks** [comment]
Windows 10 Home 21H1
Reinstall MS VIsual Studio 2019 Community from scratch
Reinstall NVIDIA cuda_11.3.1_465.89_win10.exe
git clone https://github.com/microsoft/vcpkg.git
I tried with both the master branch and the 2021.05.12 tag -> same result.
.\bootstrap-vcpkg.bat
.\vcpkg.exe integrate install
.\vcpkg.exe install cli11:x64-windows pybind11:x64-windows spdlog:x64-windows cuda:x64-windows

In the provided testcase
cd Testcase\DevOps\LocalHost
.\Generate-Packages-Windows-CommandLine.cmd

It will execute
```
rmdir /s/q Generate-Packages-Windows-CommandLine
mkdir      Generate-Packages-Windows-CommandLine
pushd      Generate-Packages-Windows-CommandLine

call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars64.bat"

cmake ^
    -G "NMake Makefiles" ^
    -D CMAKE_C_COMPILER=cl ^
    -D CMAKE_CXX_COMPILER=cl ^
    -D CMAKE_BUILD_TYPE=Release ^
    -D CMAKE_TOOLCHAIN_FILE=C:\Tools\VCPKG-2021.05.12\scripts\buildsystems\vcpkg.cmake ^
    -D PYBIND11_PYTHON_VERSION=3.7 ^
    -D EAJKS_BUILD_SDK=ON ^
    -D EAJKS_BUILD_SDK_SANDBOXES=ON ^
    ..\..\..\Components

nmake package

popd
```

The top-level CMakeLists.txt is in Testcase\Components
The CUDA subproject is in Testcase\Components\SDK\Libraries\SandBox-GPU\CMakeLists.txt

The error message
```
**********************************************************************
** Visual Studio 2019 Developer Command Prompt v16.10.2
** Copyright (c) 2021 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x64'
-- The CXX compiler identification is MSVC 19.29.30038.1
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30037/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - not found
-- Found Threads: TRUE  
CMake Error at C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.20/Modules/CMakeDetermineCUDACompiler.cmake:207 (message):
  Couldn't find CUDA library root.
Call Stack (most recent call first):
  SDK/Libraries/SandBox-GPU/CMakeLists.txt:6 (PROJECT)


-- Configuring incomplete, errors occurred!
See also "C:/by-EAjks.Com/Value Propositions/Testcase/DevOps/LocalHost/Generate-Packages-Windows-CommandLine/CMakeFiles/CMakeOutput.log".
See also "C:/by-EAjks.Com/Value Propositions/Testcase/DevOps/LocalHost/Generate-Packages-Windows-CommandLine/CMakeFiles/CMakeError.log".

Microsoft (R) Program Maintenance Utility Version 14.29.30038.1
Copyright (C) Microsoft Corporation.  All rights reserved.

NMAKE : fatal error U1073: don't know how to make 'package'
Stop.
```

When I use integrate the cli11, pybind11, spdlog in the tree, add them subprojects, and remove the vcpkg buildchain, it compiles without any problem. Only when I try to use vcpkg dependencies, there seems to be an issue.
[Testcase.zip](https://github.com/microsoft/vcpkg/files/6662206/Testcase.zip)

**Comment by NancyLi1013** [comment]
@EajksEajks 

I cannot reproduce this issue on my local with the above testcase you attached. Could you please check if you have added `CUDA` path in your environment? 

```
PS F:\Testcase\DevOps\LocalHost> .\Generate-Packages-Windows-CommandLine.cmd
**********************************************************************
** Visual Studio 2019 Developer Command Prompt v16.10.2
** Copyright (c) 2021 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x64'
-- The CXX compiler identification is MSVC 19.29.30038.1
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.29.30037/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - not found
-- Found Threads: TRUE
-- The CUDA compiler identification is NVIDIA 11.3.58
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Check for working CUDA compiler: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.3/bin/nvcc.exe - skipped
-- Detecting CUDA compile features
-- Detecting CUDA compile features - done
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_C_COMPILER
    PYBIND11_PYTHON_VERSION


-- Build files have been written to: F:/Testcase/DevOps/LocalHost/Generate-Packages-Windows-CommandLine
```
